docs/doxygen/man/man3/SendOne.3
.TH "SendOne" 3 "Thu Aug 30 2018" "esc" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SendOne \- 
.SH SYNOPSIS
.br
.PP
.PP
Inherits \fBBlockCommand\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSendOne\fP (uint16_t abank, uint32_t auser, uint32_t amsid, uint16_t bbank, uint32_t buser, int64_t tmass, uint8_t tinfo[32], uint32_t time)"
.br
.ti -1c
.RI "virtual int \fBgetType\fP () override"
.br
.RI "\fIReturn TXSTYPE_PUT as type \&. \fP"
.ti -1c
.RI "virtual CommandType \fBgetCommandType\fP () override"
.br
.RI "\fIReturn eModifying as command type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetData\fP () override"
.br
.RI "\fIGet pointer to command data structure\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetResponse\fP () override"
.br
.RI "\fIGet pointer to response data\&. \fP"
.ti -1c
.RI "virtual void \fBsetData\fP (char *data) override"
.br
.RI "\fIPut data as a char list and put convert it to data structure\&. \fP"
.ti -1c
.RI "virtual void \fBsetResponse\fP (char *response) override"
.br
.RI "\fIApply data as a response struct\&. \fP"
.ti -1c
.RI "virtual int \fBgetDataSize\fP () override"
.br
.RI "\fIGet data struct size\&. Without signature\&. \fP"
.ti -1c
.RI "virtual int \fBgetResponseSize\fP () override"
.br
.RI "\fIGet response data struct size\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetSignature\fP () override"
.br
.RI "\fIGet pointer to signature data\&. \fP"
.ti -1c
.RI "virtual int \fBgetSignatureSize\fP () override"
.br
.RI "\fIGet signature size\&. \fP"
.ti -1c
.RI "virtual void \fBsign\fP (const uint8_t *hash, const uint8_t *sk, const uint8_t *pk) override"
.br
.RI "\fISign actual data plus hash using user private and public keys\&. \fP"
.ti -1c
.RI "virtual bool \fBcheckSignature\fP (const uint8_t *hash, const uint8_t *pk) override"
.br
.RI "\fICheck actual signature\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetTime\fP () override"
.br
.RI "\fIGet time of command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserId\fP () override"
.br
.RI "\fIGet User ID\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetBankId\fP () override"
.br
.RI "\fIGet Node ID\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetFee\fP () override"
.br
.RI "\fIGet command fee\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetDeduct\fP () override"
.br
.RI "\fIGet change in cash balance after command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserMessageId\fP () override"
.br
.RI "\fIGet user message id\&. \fP"
.ti -1c
.RI "virtual uint8_t * \fBgetInfoMsg\fP ()"
.br
.RI "\fIGet additional info typed by client in transaction message\&. \fP"
.ti -1c
.RI "virtual bool \fBsend\fP (\fBINetworkClient\fP &netClient) override"
.br
.RI "\fISend data to the server\&. \fP"
.ti -1c
.RI "virtual void \fBsaveResponse\fP (\fBsettings\fP &sts) override"
.br
.RI "\fISave command response to settings object\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetBlockMessage\fP () override"
.br
.RI "\fIGet data as it is stored in blockchain\&. \fP"
.ti -1c
.RI "virtual size_t \fBgetBlockMessageSize\fP () override"
.br
.RI "\fIGet blockchain data size\&. \fP"
.ti -1c
.RI "virtual std::string \fBtoString\fP (bool pretty) override"
.br
.ti -1c
.RI "virtual void \fBtoJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual void \fBtxnToJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual std::string \fBusageHelperToString\fP () override"
.br
.ti -1c
.RI "virtual uint32_t \fBgetDestBankId\fP ()"
.br
.RI "\fIGet destination bank id\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetDestUserId\fP ()"
.br
.RI "\fIGet destination user id\&. \fP"
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fBUserSendOne\fP \fBm_data\fP"
.br
.ti -1c
.RI "\fBcommandresponse\fP \fBm_response\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 8 of file sendone\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "bool SendOne::checkSignature (const uint8_t * hash, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Check actual signature\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 62 of file sendone\&.cpp\&.
.SS "bool SendOne::send (\fBINetworkClient\fP & netClient)\fC [override]\fP, \fC [virtual]\fP"

.PP
Send data to the server\&. 
.PP
\fBParameters:\fP
.RS 4
\fInetClient\fP Netwrok client implementation of \fBINetworkClient\fP interface\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 119 of file sendone\&.cpp\&.
.SS "void SendOne::sign (const uint8_t * hash, const uint8_t * sk, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Sign actual data plus hash using user private and public keys\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIsk\fP Pointer to provate key\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 58 of file sendone\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for esc from the source code\&.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-24
### Changed
- Genesis file for the launch
- Update README

## [0.0.6] - 2018-08-10
### Added
- Profit sharing among TOP and VIP nodes

### Changed
- Genesis file for test launch

### Fixed
- Vote counting in get_block command
- Ability to send externally signed transactions

## [0.0.5] - 2018-08-01
### Added
- Support passing version by cmake command.
- Contributing guide
- "Dry run" test case

### Fixed
- Node stops after incorrect `send_many` request  - #134
- Fee checking
- Internal transactions numbers

## [0.0.4] - 2018-07-27
### Changed
- Update default parameters
- Creating a dev version with reference to the last tag

[Unreleased]: https://github.com/adshares/ads/compare/v1.0.0...HEAD

[0.0.5]: https://github.com/adshares/ads/compare/v0.0.6...v1.0.0
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/adshares/ads/compare/v0.0.3...v0.0.4


src/common/command/sendone.cpp
#include "sendone.h"
#include "ed25519/ed25519.h"
#include "abstraction/interfaces.h"
#include "helper/json.h"
#include "helper/hash.h"

SendOne::SendOne()
    : m_data{} {
    m_responseError = ErrorCodes::Code::eNone;
}

SendOne::SendOne(uint16_t abank, uint32_t auser, uint32_t amsid, uint16_t bbank,
                 uint32_t buser, int64_t tmass, uint8_t tinfo[32], uint32_t time)
    : m_data(abank, auser, amsid, bbank, buser, tmass, time, tinfo) {
    m_responseError = ErrorCodes::Code::eNone;
}

unsigned char* SendOne::getData() {
    return reinterpret_cast<unsigned char*>(&m_data.info);
}

unsigned char* SendOne::getResponse() {
    return reinterpret_cast<unsigned char*>(&m_response);
}

void SendOne::setData(char* data) {
    m_data = *reinterpret_cast<decltype(m_data)*>(data);
}

void SendOne::setResponse(char* response) {
    m_response = *reinterpret_cast<decltype(m_response)*>(response);
}

int SendOne::getDataSize() {
    return sizeof(m_data.info);
}

int SendOne::getResponseSize() {
    return sizeof(m_response);
}

unsigned char* SendOne::getSignature() {
    return m_data.sign;
}

int SendOne::getSignatureSize() {
    return sizeof(m_data.sign);
}

int SendOne::getType() {
    return TXSTYPE_PUT;
}

CommandType SendOne::getCommandType() {
    return CommandType::eModifying;
}

void SendOne::sign(const uint8_t* hash, const uint8_t* sk, const uint8_t* pk) {
    ed25519_sign2(hash,SHA256_DIGEST_LENGTH , getData(), getDataSize(), sk, pk, getSignature());
}

bool SendOne::checkSignature(const uint8_t* hash, const uint8_t* pk) {
    return (ed25519_sign_open2(hash, SHA256_DIGEST_LENGTH,getData(),getDataSize(),pk,getSignature()) == 0);

}

void SendOne::saveResponse(settings& sts) {
    if (!sts.without_secret && !std::equal(sts.pk, sts.pk + SHA256_DIGEST_LENGTH, m_response.usera.pkey)) {
        m_responseError = ErrorCodes::Code::ePkeyDiffers;
    }

    std::array<uint8_t, SHA256_DIGEST_LENGTH> hashout;
    Helper::create256signhash(getSignature(), getSignatureSize(), sts.ha, hashout);
    if (!sts.signature_provided && !std::equal(hashout.begin(), hashout.end(), m_response.usera.hash)) {
        m_responseError = ErrorCodes::Code::eHashMismatch;
    }

    sts.msid = m_response.usera.msid;
    std::copy(m_response.usera.hash, m_response.usera.hash + SHA256_DIGEST_LENGTH, sts.ha.data());
}

uint32_t SendOne::getUserId() {
    return m_data.info.auser;
}

uint32_t SendOne::getBankId() {
    return m_data.info.abank;
}

uint32_t SendOne::getTime() {
    return m_data.info.ttime;
}

unsigned char*  SendOne::getBlockMessage()
{
    return reinterpret_cast<unsigned char*>(&m_data);
}

size_t  SendOne::getBlockMessageSize()
{
    return sizeof(UserSendOne);
}

int64_t SendOne::getFee() {
    int64_t fee=TXS_PUT_FEE(m_data.info.ntmass);
    if(m_data.info.abank!=m_data.info.bbank) {
        fee += TXS_LNG_FEE(m_data.info.ntmass);
    }
    if (fee < TXS_MIN_FEE) {
        fee = TXS_MIN_FEE;
    }
    return fee;
}

int64_t SendOne::getDeduct() {
    return m_data.info.ntmass;
}

bool SendOne::send(INetworkClient& netClient) {
    if(!netClient.sendData(getData(), sizeof(m_data))) {
        ELOG("SendOne sending error\n");
        return false;
    }

    if (!netClient.readData((int32_t*)&m_responseError, ERROR_CODE_LENGTH)) {
        ELOG("SendOne reading error\n");
        return false;
    }

    if (m_responseError) {
        return true;
    }

    if(!netClient.readData(getResponse(), getResponseSize())) {
        ELOG("SendOne ERROR reading global info\n");
        return false;
    }

    return true;
}

uint32_t SendOne::getDestBankId() {
    return m_data.info.bbank;
}

uint32_t SendOne::getDestUserId() {
    return m_data.info.buser;
}

uint32_t SendOne::getUserMessageId() {
    return m_data.info.namsid;
}

uint8_t* SendOne::getInfoMsg() {
    return m_data.info.ntinfo;
}

std::string SendOne::toString(bool /*pretty*/) {
    return "";
}

void SendOne::toJson(boost::property_tree::ptree& ptree) {
    if (!m_responseError) {
        Helper::print_user(m_response.usera, ptree, true, this->getBankId(), this->getUserId());
        Helper::print_msgid_info(ptree, m_data.info.abank, m_response.msid, m_response.mpos);
    } else {
        if (m_responseError == ErrorCodes::Code::ePkeyDiffers) {
            std::stringstream tx_user_hashin{};
            Helper::ed25519_key2text(tx_user_hashin, m_response.usera.pkey, SHA256_DIGEST_LENGTH);
            ptree.put("tx.account_public_key_new", tx_user_hashin.str());
        }
        ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(m_responseError));
    }
}

void SendOne::txnToJson(boost::property_tree::ptree& ptree) {
    using namespace Helper;
    ptree.put(TAG::TYPE, getTxnName(m_data.info.ttype));
    ptree.put(TAG::SRC_NODE, m_data.info.abank);
    ptree.put(TAG::SRC_USER, m_data.info.auser);
    ptree.put(TAG::MSGID, m_data.info.namsid);
    ptree.put(TAG::TIME, m_data.info.ttime);
    ptree.put(TAG::DST_NODE, m_data.info.bbank);
    ptree.put(TAG::DST_USER, m_data.info.buser);
    ptree.put(TAG::SENDER_FEE, print_amount(this->getFee()));
    ptree.put(TAG::SRC_ADDRESS, Helper::print_address(m_data.info.abank, m_data.info.auser));
    ptree.put(TAG::DST_ADDRESS, Helper::print_address(m_data.info.bbank, m_data.info.buser));
    ptree.put(TAG::AMOUNT, print_amount(m_data.info.ntmass));
    ptree.put(TAG::MSG, Helper::ed25519_key2text(m_data.info.ntinfo, sizeof(m_data.info.ntinfo)));
    ptree.put(TAG::SIGN, ed25519_key2text(getSignature(), getSignatureSize()));
}

std::string SendOne::usageHelperToString() {
    std::stringstream ss{};
    ss << "Usage: " << "{\"run\":\"send_one\",\"address\":<destination_account_id>,\"amount\":<sending_amount>, [\"message\":\"message_in_hex\"]}" << "\n";
    ss << "Example: " << "{\"run\":\"send_one\",\"address\":\"0003-00000000-XXXX\",\"amount\":2.1,\"message\":\"000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F\"}" << "\n";
    return ss.str();
}
